keps/2076-kueuectl/README.md

```diff
@@ -0,0 +1,342 @@
+
+### Non-Goals
+
+* Provide any other interface than a command line (no web ui).
+* Provide a tool targeted at ml researchers to make running jobs on Kubernetes easier.
+* Expose additional metrics or statuses.
+
+## Proposal
+
+Create kueue kubectl plugin with a set of listing and management commands in form of:
+
+```
+kubectl kueue <command> <object> <flags>
+```
+
+Additionally provide a wrapper script to allow shorter syntax like:
+
+```
+kueuectl <command> <object> <flags>
+```
+
+The commands automatically submit all changes unless `--dry-run` option is given - in that 
+case the tool will print out yamls without making any changes in the cluster.
+
+### User Stories (Optional)
+
+#### Story 1
+
+I want to stop admission of new Workloads on a specific ClusterQueue but allow the already
+running Workloads to complete without manually editing the CQ's definition.
+
+#### Story 2
+
+I want to create a LocalQueue pointing to a specific CQ without creating a one-time yaml.
+
+### Risks and Mitigations
+
+* There will be an additional binary placed on sysadmin's machine, whose version will
+have to be keep in sync.
+
+## Design Details
+
+The following commands will be provided within the plugin:
+
+### Create ClusterQueue 
+
+Creates a ClusterQueue with the given name, cohort, specified quota and other details.
+Format:
+
+```
+kueuectl create cq|clusterqueue cqname 
+–-cohort=cohortname                        # defaults to "" - no cohort
+
+--queuing-strategy=strategy                # defaults to BestEffortFIFO
+--namespace-selector=selector              # defaults to {} - all namespaces can use the queue
+--reclaim-within-cohort=policy             # defaults to Never
+--preemption-within-cluster-queue = policy # defaults to Never
+--preemption-when-borrowing = policy       # defaults to Never
+
+–-nominal-quota=rfname1:resource1=value,resource2=value,resource3=value
+–-borrowing-limit=rfname1:resource1=value,resource2=value,resource3=value
+–-lending-limit=rfname1:resource1=value,resource2=value,resource3=value
+```
+
+It is possible to create a ClusterQueue with multiple resource flavors/FlavorQuotas inside
+a ResourceGroup and multiple ResourceGroups covering different sets of resources.
+The command will create the appropriate resource groups with resource flavors 
+in the order they appear in the command line. If two settings have at least 
+one common resource quota specified, they will land in the same ResourceGroup.
+
+Output: 
+A simple confirmation, like in regular kubectl create, 
+`clusterqueue.kueue.x-k8s.io/xxxxx created`
+
+
+### Create LocalQueue
+
+Creates a LocalQueue with the given name pointing to specified ClusterQueue.
+The command validates that the target ClusterQueue exists and 
+its namespace selector matches to LocalQueue's namespace.
+
+Format:
+
+```
+kueuectl create lq|localqueue lqname
+–-namespace=namespace                  # uses context's default namespace if not specified
```

Comment:
I'm not sure if it is possible, but couldn't get namespace from root kubectl arg?

```
$ kubectl --namespace=foo kueue ...
```

Reply:
@mbobrovskyi is writing the first command. Did you try putting `-n` before `kueue`?

Reply:
Nope. 

`Error: flags cannot be placed before plugin name: --namespace=foo`

Reply:
I guess that's a limitation of the plugin framework.

Reply:
Yes, but we can put it here `kubectl kueue --namespace=foo create lq lq1 -c cq1`